File: content/2-sorting/2-quick-sort/py/QuickSort.py
def quick_sort(arr):
    def partition(arr, low, high):
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        arr[i+1], arr[high] = arr[high], arr[i+1]
        return i+1

    def quick_sort_rec(arr, low, high):
        if low < high:
            pi = partition(arr, low, high)
            quick_sort_rec(arr, low, pi-1)
            quick_sort_rec(arr, pi+1, high)

    quick_sort_rec(arr, 0, len(arr)-1)
    return arr

File: content/2-sorting/2-quick-sort/rs/quick_sort.rs
fn quick_sort(arr: &mut [i32], low: isize, high: isize) {
    if low < high {
        let pi = partition(arr, low, high);
        quick_sort(arr, low, pi - 1);
        quick_sort(arr, pi + 1, high);
    }
}

fn partition(arr: &mut [i32], low: isize, high: isize) -> isize {
    let pivot = arr[high as usize];
    let mut i = low - 1;

    for j in low..high {
        if arr[j as usize] < pivot {
            i += 1;
            arr.swap(i as usize, j as usize);
        }
    }
    arr.swap((i + 1) as usize, high as usize);
    i + 1
}